src/main/java/org/eclipse/yasson/internal/Marshaller.java

```diff
@@ -35,10 +37,16 @@
 /**
  * JSONB marshaller. Created each time marshalling operation called.
  */
-public class Marshaller extends ProcessingContext implements SerializationContext {
+public class Marshaller implements SerializationContext {
 
     private static final Logger LOGGER = Logger.getLogger(Marshaller.class.getName());
 
+    /**
+     * Used to avoid StackOverflowError, when adapted / serialized object
+     * contains contains instance of its type inside it or when object has recursive reference.
+     */
+    private final List<Object> currentlyProcessedObjects = new ArrayList<>();
```

Comment:
I think we should still use `Set` instead of `List` here because `addProcessedObject` does a `contains()` which is O(N) for lists and O(1) for a HashSet. Performance may be better with List on the few benchmarks we have, but I could see it being an issue for some larger data sets

Reply:
That makes sense. :) Changing back to a HashSet